File: app/src/main/java/com/bignerdranch/android/valdueza_rpsgame/MainActivity.java
package com.bignerdranch.android.valdueza_rpsgame;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.os.CountDownTimer;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    private Button rockBtn, paperBtn, scissorsBtn,
            bestOf3btn, bestOf5Btn, bestOf7btn;
    private int turns = 0;
    private int round = 1;
    private int player2choice;
    private TextView messageTextView;
    private int roundScore = 0;
    private String p2StrChoice = "";
    private String roundResult = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        rockBtn = findViewById(R.id.rock_btn);
        paperBtn = findViewById(R.id.paper_btn);
        scissorsBtn = findViewById(R.id.scissors_btn);
        bestOf3btn = findViewById(R.id.bestOf3_btn);
        bestOf5Btn = findViewById(R.id.bestOf5_btn);
        bestOf7btn = findViewById(R.id.bestOf7_btn);

        messageTextView = findViewById(R.id.message_here);
        resetMessageText();
        activeBtnRockPaperSci(false);

        bestOf3btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                activeBtnRockPaperSci(true);
                activeBtnBestOf(false);
                turns = 3;
            }
        });

        rockBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                player2Rolled();
                effectRPS("rock");
                roundResult = getRoundResult(0);
                messageTextView.setText("Player 2 - " +
                        p2StrChoice + "\nRound " +
                        roundResult + "\n" + round + "\\" + turns);
                if(roundScore > (turns - roundScore)){
                    messageTextView.setText("Congrats. You win!");
                }
            }
        });

        paperBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                player2Rolled();
                effectRPS("paper");
                if(player2choice == 0){
                    showMsg("Player 2 picked rock! You win a round");
                } else if (player2choice == 1){
                    showMsg("Player 2 picked paper! It's a draw");
                } else {
                    showMsg("Player 2 picked scissors! You lose a round");
                }
            }
        });

        scissorsBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                player2Rolled();
                effectRPS("scissors");
                if(player2choice == 0){
                    showMsg("Player 2 picked rock! You lose a round");
                } else if (player2choice == 1){
                    showMsg("Player 2 picked paper! You win a round");
                } else {
                    showMsg("Player 2 picked scissors! It's a draw");
                }
            }
        });

    }

    private String getRoundResult(int player1){
        if(player1 == player2choice){
            return "Draw";
        } else if (player1 == 0 && player2choice == 1 ||
                    player1 == 1 && player2choice == 2 ||
                    player1 == 2 && player2choice == 0){
            return "Loss";
        } else {
            roundScore++;
            return "Won";
        }
    }

    private void activeBtnRockPaperSci(boolean enabled){
        rockBtn.setEnabled(enabled);
        paperBtn.setEnabled(enabled);
        scissorsBtn.setEnabled(enabled);
    }

    private void activeBtnBestOf(boolean enabled){
        bestOf3btn.setEnabled(enabled);
        bestOf5Btn.setEnabled(enabled);
        bestOf7btn.setEnabled(enabled);
    }

    private void player2Rolled(){
        player2choice = ((int) (Math.random()*(3)));
        if(player2choice == 0){
            p2StrChoice = "Rock";
        } else if (player2choice == 1){
            p2StrChoice = "Paper";
        } else {
            p2StrChoice = "Scissors";
        }
    }

    private void showMsg(String msg){
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    private void effectRPS(String btn){
        if(btn == "rock"){
            rockBtn.setEnabled(false);
        } else if (btn == "paper"){
            paperBtn.setEnabled(false);
        } else {
            scissorsBtn.setEnabled(false);
        }
        rockBtn.setClickable(false);
        paperBtn.setClickable(false);
        scissorsBtn.setClickable(false);
        new CountDownTimer(2000, 1000){
            @Override
            public void onTick(long l) {

            }
            public void onFinish(){
                paperBtn.setClickable(true);
                scissorsBtn.setClickable(true);
                rockBtn.setClickable(true);
                paperBtn.setEnabled(true);
                scissorsBtn.setEnabled(true);
                rockBtn.setEnabled(true);
            }
        }.start();
    }

    private void resetMessageText(){
        messageTextView.setText("Player 1 - You \nPlayer 2 - Computer \nPick a \"Best of\" category");
    }
}

File: app/src/main/java/com/bignerdranch/android/valdueza_rpsgame/MainActivity.kt
package com.bignerdranch.android.valdueza_rpsgame

import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import android.os.CountDownTimer
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*


class MainActivity : AppCompatActivity() {
    private var turns = 0
    private var round = 1
    private var player2choice = 0
    private var p1RoundScore = 0
    private var p2RoundScore = 0
    private var wins = 0
    private var p2StrChoice = ""
    private var roundResult = ""
    private var roundComplete = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        resetMessageText()
        activeBtnRockPaperSci(false)
        p1_btn.isClickable = false
        p2_btn.isClickable = false
        p1_btn.setBackgroundColor(Color.parseColor("#2643b5"))
        p2_btn.setBackgroundColor(Color.parseColor("#b5ae26"))
        bestOf3_btn?.setOnClickListener(View.OnClickListener {
            message_here.text = "Best of 3\nPlay your hand"
            activeBtnRockPaperSci(true)
            activeBtnBestOf(false)
            resetAll()
            turns = 3
            layout_main.setBackgroundColor(Color.parseColor("#FFFFFF"))
        })
        bestOf5_btn.setOnClickListener(View.OnClickListener {
            message_here?.text = "Best of 5\nPlay your hand"
            activeBtnRockPaperSci(true)
            activeBtnBestOf(false)
            resetAll()
            turns = 5
            layout_main.setBackgroundColor(Color.parseColor("#FFFFFF"))
        })
        bestOf7_btn.setOnClickListener(View.OnClickListener {
            message_here?.text = "Best of 7\nPlay your hand"
            activeBtnRockPaperSci(true)
            activeBtnBestOf(false)
            resetAll()
            turns = 7
            layout_main.setBackgroundColor(Color.parseColor("#FFFFFF"))
        })
        rock_btn.setOnClickListener(View.OnClickListener {
            player2Rolled()
            updateRound(0, "rock")
        })
        paper_btn.setOnClickListener(View.OnClickListener {
            player2Rolled()
            updateRound(1, "paper")
        })
        scissors_btn.setOnClickListener(View.OnClickListener {
            player2Rolled()
            updateRound(2, "scissors")
        })
    }

    private fun updateRound(btnPosition: Int, btnPressed: String) {
        roundResult = getRoundResult(btnPosition)
        val strOpeningMsg = """
            Player 2 - $p2StrChoice
            Round $roundResult
            $round\$turns
            """.trimIndent()
        val strWinMsg = """
            Player 2 - $p2StrChoice
            Congrats. You win!
            """.trimIndent()
        val strLostMessage = """
            Player 2 - $p2StrChoice
            You Lost the game.
            """.trimIndent()
        val strDrawMessage = """
            Player 2 - $p2StrChoice
            It's a DRAW. No one wins
            """.trimIndent()
        val strP1RoundScore = "P1: $p1RoundScore"
        val strP2RoundScore = "P2: $p2RoundScore"
        message_here!!.text = strOpeningMsg
        if (p1RoundScore > turns - p1RoundScore ||
            round == turns && p2RoundScore < p1RoundScore
        ) {
            wins++
            wins_text!!.text = "Player 1:\n$wins Wins"
            layout_main!!.setBackgroundColor(Color.parseColor("#7aeb34"))
            roundComplete = true
            message_here!!.text = strWinMsg
            activeBtnRockPaperSci(false)
            activeBtnBestOf(true)
        } else if (p2RoundScore > turns - p2RoundScore ||
            round == turns && p2RoundScore > p1RoundScore
        ) {
            message_here!!.text = strLostMessage
            activeBtnRockPaperSci(false)
            activeBtnBestOf(true)
            layout_main!!.setBackgroundColor(Color.parseColor("#eb4034"))
            roundComplete = true
        } else if (round == turns) {
            message_here!!.text = strDrawMessage
            activeBtnRockPaperSci(false)
            activeBtnBestOf(true)
            layout_main!!.setBackgroundColor(Color.parseColor("#a0b9e8"))
            roundComplete = true
        }
        effectRPS(btnPressed)
        p1_btn!!.text = strP1RoundScore
        p2_btn!!.text = strP2RoundScore
        round++
    }

    private fun resetAll() {
        round = 1
        p1RoundScore = 0
        p2RoundScore = 0
        roundComplete = false
        p1_btn!!.text = "P1"
        p2_btn!!.text = "P2"
    }

    private fun getRoundResult(player1: Int): String {
        return if (player1 == player2choice) {
            "Draw"
        } else if (player1 == 0 && player2choice == 1 || player1 == 1 && player2choice == 2 || player1 == 2 && player2choice == 0) {
            p2RoundScore++
            "Loss"
        } else {
            p1RoundScore++
            "Won"
        }
    }

    private fun activeBtnRockPaperSci(enabled: Boolean) {
        rock_btn!!.isEnabled = enabled
        paper_btn!!.isEnabled = enabled
        scissors_btn!!.isEnabled = enabled
        rock_btn!!.isClickable = enabled
        paper_btn!!.isClickable = enabled
        scissors_btn!!.isClickable = enabled
    }

    private fun activeBtnBestOf(enabled: Boolean) {
        bestOf3_btn!!.isEnabled = enabled
        bestOf5_btn!!.isEnabled = enabled
        bestOf7_btn!!.isEnabled = enabled
    }

    private fun player2Rolled() {
        player2choice = (Math.random() * 3).toInt()
        p2StrChoice = if (player2choice == 0) {
            "Rock"
        } else if (player2choice == 1) {
            "Paper"
        } else {
            "Scissors"
        }
    }

    private fun showMsg(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun effectRPS(btn: String) {
        if (btn == "rock") {
            rock_btn!!.isEnabled = false
        } else if (btn == "paper") {
            paper_btn!!.isEnabled = false
        } else {
            scissors_btn!!.isEnabled = false
        }
        rock_btn!!.isClickable = false
        paper_btn!!.isClickable = false
        scissors_btn!!.isClickable = false
        if (!roundComplete) {
            object : CountDownTimer(1000, 1000) {
                override fun onTick(l: Long) {}
                override fun onFinish() {
                    paper_btn!!.isClickable = true
                    scissors_btn!!.isClickable = true
                    rock_btn!!.isClickable = true
                    paper_btn!!.isEnabled = true
                    scissors_btn!!.isEnabled = true
                    rock_btn!!.isEnabled = true
                }
            }.start()
        }
    }

    private fun resetMessageText() {
        val firstMsg = """Player 1 - You 
Player 2 - Computer
Pick a "Best of" category"""
        message_here!!.text = firstMsg
    }
}